mm2src/coins/utxo/utxo_builder/utxo_arc_builder.rs

```diff
@@ -339,16 +349,30 @@ pub(crate) async fn block_header_utxo_loop<T: UtxoCommonOps>(
             },
         };
 
-        // Validate retrieved block headers
-        if let Some(params) = &coin.as_ref().conf.block_headers_verification_params {
-            if let Err(e) = validate_headers(ticker, from_block_height, block_headers, storage, params).await {
-                error!("Error {e:?} on validating the latest headers for {ticker}!");
-                // Todo: remove this electrum server and use another in this case since the headers from this server are invalid
-                sync_status_loop_handle.notify_on_permanent_error(e.to_string());
-                break;
-            }
+        // Validate retrieved block headers.
+        if let Err(e) = validate_headers(ticker, from_block_height, block_headers, storage, &spv_conf).await {
+            error!("Error {e:?} on validating the latest headers for {ticker}!");
+            // Todo: remove this electrum server and use another in this case since the headers from this server are invalid
+            sync_status_loop_handle.notify_on_permanent_error(e.to_string());
+            break;
         };
 
+        // Check if we need to max the number of headers to be stored in storage.
+        if let Some(max_stored_block_headers) = spv_conf.max_stored_block_headers {
+            let headers_in_storage_count = storage.get_total_block_headers_from_storage().await.unwrap_or(0);
+            let current_total = block_registry.len() as u64 + headers_in_storage_count;
+            let max_stored_block_headers = max_stored_block_headers.get();
+
+            if current_total >= max_stored_block_headers {
```

Comment:
shouldn't this be `current_total > max_stored_block_headers` ?

Reply:
yeah something like that but this actually 
```rust
 if current_total >= max_stored_block_headers {
                let headers_to_remove = if current_total > max_stored_block_headers {
                    (current_total - max_stored_block_headers) as i64
                } else if current_total == max_stored_block_headers {
                    max_stored_block_headers - lock_registry.len() as u64
                } else {
                    0
                };
                storage
                    .remove_block_headers_from_storage(headers_to_remove)
                    .await
                    .error_log();
            }
```

Reply:
Why would we want to remove `0` headers? This will just make an unnecessary call to the DB.

Reply:
yes I'm aware that's why I have this check 
```rust
    async fn remove_block_headers_from_storage(&self, limit: i64) -> Result<(), BlockHeaderStorageError> {
        if limit < 1 {
            return Ok(());
        }
```

Reply:
but I guess this is better 
```rust
async fn calculate_and_remove_headers_from_db(
    storage: &BlockHeaderStorage,
    max_stored_block_headers: Option<NonZeroU64>,
    block_registry_len: u64,
) -> Result<(), BlockHeaderStorageError> {
    if let Some(max_stored_block_headers) = max_stored_block_headers {
        let headers_in_storage_count = storage.get_total_block_headers_from_storage().await.unwrap_or(0);
        let current_total = block_registry_len + headers_in_storage_count;
        let max_stored_block_headers = max_stored_block_headers.get();

        if current_total >= max_stored_block_headers {
            let mut headers_to_remove = 0;

            if current_total > max_stored_block_headers {
                headers_to_remove = current_total - max_stored_block_headers
            }

            if current_total == max_stored_block_headers {
                headers_to_remove = max_stored_block_headers - block_registry_len
            };

            if headers_to_remove.is_non_zero() {
                return storage
                    .remove_block_headers_from_storage(headers_to_remove as i64)
                    .await;
            }
        }
    };

    Ok(())
}
```

Reply:
https://github.com/KomodoPlatform/atomicDEX-API/pull/1585#discussion_r1081521987